bundles/org.eclipse.swt/Eclipse SWT Custom Widgets/common/org/eclipse/swt/custom/BusyIndicator.java

```diff
@@ -70,6 +70,41 @@ public static void showWhile(Display display, Runnable runnable) {
 		}
 	}
 
+	/**
+	 * If called from a {@link Display} thread, waits for the given
+	 * <code>Future</code> to complete and providing busy feedback using the
+	 * busy indicator. While waiting for completion pending UI events are processed
+	 * to prevent UI freeze.
+	 *
+	 * If there is no {@link Display} for the current thread, the method return
+	 * immediately and no feedback is shown to the user.
+	 *
+	 * In any case, the passed {@link Future} is returned for chaining of further operations.
+	 *
+	 * @param future the {@link Future} for which busy feedback is to be shown.
+	 * @return the {@link Future} for chaining.
+	 * @since 3.127
+	 */
+	public static <F extends Future<T>, T> F showWhile(F future) {
+		Objects.requireNonNull(future);
+		if (!future.isDone()) {
+			Display display = Display.findDisplay(Thread.currentThread());
+			if (display != null) {
+				Integer busyId = setBusyCursor(display);
+				try {
+					while (!future.isDone() && !display.isDisposed()) {
+						if (!display.readAndDispatch()) {
+							Thread.onSpinWait();
```

Comment:
Shouldn't we call Display.sleep()?

Reply:
`Display.sleep()` potentially waits "forever" (from the documentation):

https://github.com/eclipse-platform/eclipse.platform.swt/blob/e0356e9d745edc89d827725093aace8594da6237/bundles/org.eclipse.swt/Eclipse%20SWT/gtk/org/eclipse/swt/widgets/Display.java#L5583-L5585

the effect is that maybe the future is complete, but we never exit this method until a new event happens e.g. user move the mouse or "something" happens, so this is fine for the mainloop, but here we want to just exit as soon as possible to return control to the original eventloop (what ever that currently is...).